### PR TITLE
[improve][pip] PIP-297: Support terminating Function & Connector with the fatal exception

### PR DESCRIPTION
**Pull request by RobertIndie**
### Motivation

This PIP is to improve the current function exception handler. It will be applied to both Pulsar Function and Pulsar IO Connector.

For more details, please refer to `pip-297.md`

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [x] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pip/pip-297.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+**Function instance thread**: The function framework initializes a thread for each function instance to handle the
+core logic of the function/connector, including consuming messages from the Pulsar topic for the sink connector,
+executing the logic of the function, producing messages to the Pulsar topic for the source connector, handling the
+exception, etc.
+
+**Exception handling logic**: The function itself can throw exceptions, and this thread will catch the exception and
+then close the function. This means that the function will stop working until it is restarted manually or
+automatically by the function framework.
+
+Even though it is not explicitly defined, there are two types of exceptions that could be handled for the function or
```
Comment (codelipenghui) [inline pip/pip-297.md:34]:
```suggestion
Even though it is not explicitly defined, there are two types of exceptions that should be handled by the function or
```

File: pip/pip-297.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+the framework.
+
+- **Recoverable exception**: This is an exception that the function can recover from by itself, such as network
+  connection issues, persistence issues, etc. The function can catch these exceptions and retry the operation until it
+  succeeds or reaches a limit.
+- **Unrecoverable exception**: This is an exception that the function cannot recover from by itself and needs to notify
+  the framework to terminate it. These are fatal exceptions that indicate a configuration issue, a logic error, or an
+  incompatible system. The function framework will catch these exceptions, report them to users, and terminate the
+  function.
```
Comment (codelipenghui) [inline pip/pip-297.md:43]: Maybe Retriable and UnRetriable are more appropriate?

File: pip/pip-297.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+Introduce `raiseException` method to the `BaseContext`:
+
+```java
+public interface BaseContext {
+    /**
+     * Raise an unrecoverable exception to the function framework.
+     *
+     * @param e the exception to be raised
+     */
+    void raiseException(Exception e);
```
Comment (eolivelli) [inline pip/pip-297.md:119]: let's use "Throwable" instead of Exception

File: pip/pip-297.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+For example, suppose we have a sink connector that uses its own threads to batch-sink the data to an external system. If
+any unrecoverable exceptions occur in those threads, the function instance thread will not be aware of them and will
+not be able to terminate the connector. This will cause the connector to hang indefinitely. There is a related issue
+here: https://github.com/apache/pulsar/issues/9464
+
+The same problem exists for the source connector. The source connector may also use a separate thread to fetch data from
+an external system. If any fatal exceptions happen in that thread, the connector will also hang forever. This issue has
+been observed for the Kafka source connector: https://github.com/apache/pulsar/issues/9464. We have fixed it by adding
+the notifyError method to the PushSource class in PIP-281: https://github.com/apache/pulsar/pull/20807. However, this
+does not solve the same problem that all source connectors face.
```
Comment (codelipenghui) [inline pip/pip-297.md:59]: Why the notifyError method can't solve the same problem for all source connectors? It's not a specific fix for Kafka connector, right?

File: pip/pip-297.md (hunk 3 above)
Comment (eolivelli) [reply pip/pip-297.md:119]: I think that it is better to have two methods:
- raiseUnrecoverableError
- raiseTemporaryError

This way it is clearer for the framework what it is expected to happen, the parameter is only for descriptive purpose, the framework cannot decided if a problem is temporary or definitive given the exception (for instance an IOException may mean both temporary and permanent failures)

File: pip/pip-297.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+- The fixes of the exception-raising issue mentioned in the Motivation part for all the connectors are not included in
+  this PIP. This PIP only provides the feature for the connector developer to raise the exception outside the function
+  instance thread. The fixes should be in serval different PRs.
+
+# High Level Design
+
+Introduce a new method `raiseException` to the context. All the connector implementation code and the function code can
+use this context and call the `raiseException` method to raise an unrecoverable exception to the function framework.
+
+After the connector or function raises the unrecoverable exception, the function instance thread will be interrupted.
```
Comment (eolivelli) [inline pip/pip-297.md:88]: we must define clearly the behaviour of this method when called from threads that are not the "main" thread of the function.

File: pip/pip-297.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+been observed for the Kafka source connector: https://github.com/apache/pulsar/issues/9464. We have fixed it by adding
+the notifyError method to the PushSource class in PIP-281: https://github.com/apache/pulsar/pull/20807. However, this
+does not solve the same problem that all source connectors face.
+
+The problem is same for the Pulsar Function. The user can also process messages asynchronously in a Pulsar Function. If
+any fatal exceptions happen in the user's thread, the function framework can't handle the exception
+correctly.
+
+We need a way for the connector and function developers to throw unrecoverable exceptions outside the function instance
+thread. The function framework should catch these exceptions and terminate the function accordingly.
```
Comment (codelipenghui) [inline pip/pip-297.md:66]: How about the retriable exceptions from the user's code? Will the function framework perform the retry automatically?

File: pip/pip-297.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  this PIP. This PIP only provides the feature for the connector developer to raise the exception outside the function
+  instance thread. The fixes should be in serval different PRs.
+
+# High Level Design
+
+Introduce a new method `raiseException` to the context. All the connector implementation code and the function code can
+use this context and call the `raiseException` method to raise an unrecoverable exception to the function framework.
+
+After the connector or function raises the unrecoverable exception, the function instance thread will be interrupted.
+The function framework then could catch the exception, log it, and then terminate the function instance.
```
Comment (codelipenghui) [inline pip/pip-297.md:89]: Is it possible for users to terminate the function instance directly for now?

File: pip/pip-297.md (hunk 3 above)
Comment (codelipenghui) [inline pip/pip-297.md:119]: How does the function framework know if the exception is a retriable exception or an un-retriable exception? From the description. It looks like the connector or function should only broadcast the un-retriable exception by this method. If yes, maybe we'd better change the method name to raiseUnRetriableException().

And if there is a way to terminate the Function Instance. IMO, it should be better than raising exceptions.

File: pip/pip-297.md (hunk 6 above)
Comment (codelipenghui) [reply pip/pip-297.md:66]: Just want to understand how the connectors or functions handle exceptions.

File: pip/pip-297.md (hunk 3 above)
Comment (RobertIndie) [reply pip/pip-297.md]: I think the PIP has introduced the `Recoverable Exception` and `UnRecoverable Exception` which confused the reader. I have refined the PIP to make it clearer for the exception handling.

For the `TemporaryError` here, we don't need to notify the function framework, the function and the connector should handle it by itself.

File: pip/pip-297.md (hunk 5 above)
Comment (RobertIndie) [reply pip/pip-297.md]: I added a detailed behavior explaination to the `Design & Implementation Details` part. PTAL

File: pip/pip-297.md (hunk 4 above)
Comment (RobertIndie) [reply pip/pip-297.md]: I added the explanation to the PIP.

The PIP-281 only affects `PushSource`. Not all source connectors are implemented based on `PushSource`, such as debezium connector.

File: pip/pip-297.md (hunk 6 above)
Comment (RobertIndie) [reply pip/pip-297.md]: I just added more details on how the function framework handles exceptions in part `The current implementation of the exception handler`. There is an answer to this question. PTAL.

File: pip/pip-297.md (hunk 7 above)
Comment (RobertIndie) [reply pip/pip-297.md]: No. There is no direct way to terminate it. PIP-281 uses an indirect way to implement it by resending an exception record to the source connector.

File: pip/pip-297.md (hunk 3 above)
Comment (RobertIndie) [reply pip/pip-297.md]: Good suggestion. This method should only handle `un-retriable exception` or, to be clearer, `fatal exception`. 

I have renamed it to `raiseFatalException`.

**Review by shibd — APPROVED** [review]
/LGTM

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**